File: packages/cli/src/core/trace.ts
import fs from "fs-extra";
import * as path from "node:path";
import { createHash } from "node:crypto";

export interface TraceEvent {
  event_id: string;
  event_type: string;
  previous_hash?: string | null;
  event_hash?: string | null;
  [key: string]: unknown;
}

function sha256String(input: string): string {
  return createHash("sha256").update(input, "utf-8").digest("hex");
}

function computeEventHash(
  event: TraceEvent,
  previousHash: string | null
): string {
  // Hash canonical full event excluding previous_hash and event_hash itself.
  // previousHash (the prior event's event_hash) is included as chain linkage.
  const { previous_hash: _prev, event_hash: _h, ...rest } = event;
  const payload = {
    ...rest,
    previous_hash: previousHash,
  };
  return sha256String(JSON.stringify(payload));
}

export async function appendTrace(
  event: TraceEvent,
  traceDir = ".x-harness/traces"
): Promise<TraceEvent> {
  await fs.ensureDir(traceDir);
  const filePath = path.join(traceDir, "events.jsonl");

  // Read existing events to get the last hash
  const events = await readTrace(traceDir);
  const previousHash =
    events.length > 0 ? (events[events.length - 1].event_hash ?? null) : null;

  const enriched: TraceEvent = {
    ...event,
    previous_hash: previousHash,
    event_hash: computeEventHash(event, previousHash),
  };

  await fs.appendFile(filePath, JSON.stringify(enriched) + "\n");
  return enriched;
}

export async function readTrace(
  traceDir = ".x-harness/traces"
): Promise<TraceEvent[]> {
  const filePath = path.join(traceDir, "events.jsonl");
  return readTraceFromFile(filePath);
}

export async function readTraceFromFile(
  filePath: string
): Promise<TraceEvent[]> {
  if (!(await fs.pathExists(filePath))) return [];
  const content = await fs.readFile(filePath, "utf-8");
  return content
    .split("\n")
    .map((line: string) => line.trim())
    .filter((line: string) => line.length > 0)
    .map((line: string) => JSON.parse(line) as TraceEvent);
}

export interface ChainVerificationResult {
  valid: boolean;
  eventsChecked: number;
  firstBrokenIndex: number | null;
  firstBrokenEventId: string | null;
  expectedHash: string | null;
  actualHash: string | null;
}

export function verifyTraceChain(
  events: TraceEvent[]
): ChainVerificationResult {
  if (events.length === 0) {
    return {
      valid: true,
      eventsChecked: 0,
      firstBrokenIndex: null,
      firstBrokenEventId: null,
      expectedHash: null,
      actualHash: null,
    };
  }

  for (let i = 0; i < events.length; i++) {
    const event = events[i];
    const previousHash = i > 0 ? (events[i - 1].event_hash ?? null) : null;

    // Legacy events without event_hash are skipped in chain verification
    // but their presence is noted by the fact that we still check subsequent links
    if (!event.event_hash) {
      continue;
    }

    const expectedHash = computeEventHash(event, previousHash);
    if (event.event_hash !== expectedHash) {
      return {
        valid: false,
        eventsChecked: i + 1,
        firstBrokenIndex: i,
        firstBrokenEventId: event.event_id,
        expectedHash,
        actualHash: event.event_hash,
      };
    }

    // Also verify previous_hash linkage (for i > 0)
    if (i > 0) {
      const expectedPreviousHash = events[i - 1].event_hash ?? null;
      const actualPreviousHash = event.previous_hash ?? null;
      if (actualPreviousHash !== expectedPreviousHash) {
        return {
          valid: false,
          eventsChecked: i + 1,
          firstBrokenIndex: i,
          firstBrokenEventId: event.event_id,
          expectedHash: expectedPreviousHash,
          actualHash: actualPreviousHash,
        };
      }
    }
  }

  return {
    valid: true,
    eventsChecked: events.length,
    firstBrokenIndex: null,
    firstBrokenEventId: null,
    expectedHash: null,
    actualHash: null,
  };
}

File: internal/trace/trace.go
package trace

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// TraceEvent is a single trace event stored as a flexible map.
type TraceEvent map[string]interface{}

func (e TraceEvent) getString(key string) string {
	if v, ok := e[key]; ok {
		if s, ok := v.(string); ok {
			return s
		}
	}
	return ""
}

func sha256String(input string) string {
	h := sha256.New()
	h.Write([]byte(input))
	return fmt.Sprintf("%x", h.Sum(nil))
}

func computeEventHash(event TraceEvent, previousHash string) string {
	data, _ := json.Marshal(event)
	var m map[string]interface{}
	_ = json.Unmarshal(data, &m)
	delete(m, "previous_hash")
	delete(m, "event_hash")
	m["previous_hash"] = previousHash
	canonical, _ := json.Marshal(m)
	return sha256String(string(canonical))
}

// ReadTrace reads events from traceDir/events.jsonl.
func ReadTrace(traceDir string) ([]TraceEvent, error) {
	filePath := filepath.Join(traceDir, "events.jsonl")
	return ReadTraceFromFile(filePath)
}

// ReadTraceFromFile reads events from a specific JSONL file.
func ReadTraceFromFile(filePath string) ([]TraceEvent, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	lines := strings.Split(string(data), "\n")
	var events []TraceEvent
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		var event TraceEvent
		if err := json.Unmarshal([]byte(line), &event); err != nil {
			return nil, err
		}
		events = append(events, event)
	}
	return events, nil
}

// ChainVerificationResult is the result of VerifyTraceChain.
type ChainVerificationResult struct {
	Valid              bool   `json:"valid"`
	EventsChecked      int    `json:"events_checked"`
	FirstBrokenIndex   *int   `json:"first_broken_index,omitempty"`
	FirstBrokenEventID string `json:"first_broken_event_id,omitempty"`
	ExpectedHash       string `json:"expected_hash,omitempty"`
	ActualHash         string `json:"actual_hash,omitempty"`
}

// VerifyTraceChain verifies the integrity of the trace hash chain.
func VerifyTraceChain(events []TraceEvent) ChainVerificationResult {
	if len(events) == 0 {
		return ChainVerificationResult{Valid: true, EventsChecked: 0}
	}

	for i := 0; i < len(events); i++ {
		event := events[i]
		var previousHash string
		if i > 0 {
			previousHash = events[i-1].getString("event_hash")
		}

		// Skip legacy events without event_hash.
		if event.getString("event_hash") == "" {
			continue
		}

		expectedHash := computeEventHash(event, previousHash)
		if event.getString("event_hash") != expectedHash {
			idx := i
			return ChainVerificationResult{
				Valid:              false,
				EventsChecked:      i + 1,
				FirstBrokenIndex:   &idx,
				FirstBrokenEventID: event.getString("event_id"),
				ExpectedHash:       expectedHash,
				ActualHash:         event.getString("event_hash"),
			}
		}

		// Verify previous_hash linkage for i > 0.
		if i > 0 {
			expectedPrev := events[i-1].getString("event_hash")
			actualPrev := event.getString("previous_hash")
			if actualPrev != expectedPrev {
				idx := i
				return ChainVerificationResult{
					Valid:              false,
					EventsChecked:      i + 1,
					FirstBrokenIndex:   &idx,
					FirstBrokenEventID: event.getString("event_id"),
					ExpectedHash:       expectedPrev,
					ActualHash:         actualPrev,
				}
			}
		}
	}

	return ChainVerificationResult{
		Valid:         true,
		EventsChecked: len(events),
	}
}
